### ISSUE TITLE
Hit Breakpoints when Module is loaded later

### ISSUE THREAD

**Issue by apazureck** [Microsoft/vscode-chrome-debug#368]
- VS Code Version:
1.9.1

Hi,

Is it possible to get the chrome debugging plugin to work when the javascript file is requested via ajax/require?

I always have to add my scripts to my index.html file by hand using 
<script src="myscript.js" ></script>
in order to get my breakpoints to be hit in vscode.

myscript.js is later loaded via ajax. the internal chrome debugger is able to hit the breakpoints when I set them. In vscode the breakpoints do not get hit.

![grafik](https://cloud.githubusercontent.com/assets/14858596/23212237/1018f4a0-f906-11e6-8c15-0138cff79d2f.png)

In the image you can see the scripts in the orange folder were loaded later on.

**Comment by roblourens** [comment]
It shouldn't matter - most likely the script isn't resolved to your source code on disk correctly. You can try using the `.scripts` command in the vscode debug console to figure out whether they have been resolved correctly.

**Comment by apazureck** [comment]
Hi,

Thanks for the quick reply.

I used the scripts command, but they do not show up there. These are the loaded scripts:

```
› eval://129
› eval://263
› eval://264
› eval://308
› eval://31
› eval://320
› eval://350
› eval://361
› eval://365
› eval://370
› http://localhost:8080/ifm.datalink.linerecorder.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\ifm.datalink.linerecorder.js)
    - ifm.datalink.linerecorder.ts (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\ifm.datalink.linerecorder.ts)
› http://localhost:8080/index.html (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\index.html)
› http://localhost:8080/resources/jquery.sap.act-dbg.js?eval (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\resources\jquery.sap.act-dbg.js)

... and so on, lot of sap ui stuff ...

› http://localhost:8080/resources/sap/ui/unified/library-dbg.js?eval (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\resources\sap\ui\unified\library-dbg.js)
› http://localhost:8080/scripts/backend/communication.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\backend\communication.js)
    - communication.ts (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\backend\communication.ts)
› http://localhost:8080/scripts/backend/odataproxybase.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\backend\odataproxybase.js)
    - odataproxybase.ts (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\backend\odataproxybase.ts)
› http://localhost:8080/scripts/baseEvent.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\baseEvent.js)
    - baseEvent.ts (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\baseEvent.ts)
› http://localhost:8080/scripts/dateformat.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\dateformat.js)
› http://localhost:8080/scripts/events.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\events.js)
    - events.ts (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\events.ts)
› http://localhost:8080/scripts/linq.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\linq.js)
    - linq.ts (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\linq.ts)
› http://localhost:8080/scripts/odatajs-4.0.0.js (c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd\scripts\odatajs-4.0.0.js)
```

Here is my chrome output in my second tab. The Chrome devtools seem to get it, but it is not mapped in the debugger plugin:

![grafik](https://cloud.githubusercontent.com/assets/14858596/23253386/9a0dcfd8-f9b4-11e6-9bc2-99e40625f431.png)

And here is my launch config:

        {
            "name": "Chrome with sourcemaps",
            "type": "chrome",
            "request": "launch",
            "url": "http://localhost:8080/index.html",
            "sourceMaps": true,
            "webRoot": "${workspaceRoot}"
        }

As I read the documentation a little bit longer: may it be that the url plays along with that?

When I want to start my Hosts controller, the URL is http://localhost:8080/index.html#/hosts

**Comment by roblourens** [comment]
The stuff that's printed out by `.scripts` includes everything that Chrome has told us about up to that point in time, so if it's not there, something is wrong. Which .js file are those .ts files built into? Do you see that file in the `.scripts` output?

**Comment by apazureck** [comment]
If I get your question right:

The scripts above are all I got via .scripts. I just skipped the whole openui5 section, as there are a few. The stuff in the /scripts folder I added in the index.html. The files I am looking for are ending with .controller.js, which are not on the list, for example Agents.controller.js.

**Comment by roblourens** [comment]
Ok, they would appear to not be loaded then (or have been unloaded) in the tab you're attached to. Either you're attached to the wrong tab or something else is wrong. You can try to enable logging `"diagnosticLogging": true` then read the verbose log at the path it prints, and look for a ScriptParsed event with the script you're expecting.

**Comment by apazureck** [comment]
Here is the whole file:

If you search for hosts.controller.js you'll see the debugger seems to get the file, but my breakpiont does not get hit. I added my Agents.controller.js to my index.html and it gets hit just fine. I hope you can make more sense out of this log than I do. [Here is the whole file](https://rushfiles.one/client/publiclink.aspx?id=DLIs5h). Thanks for your help.

```
Verbose logs are written to C:\Users\pazureck\.vscode\extensions\msjsdiag.debugger-for-chrome-2.5.4\vscode-chrome-debug.txt
16:48:37, 27.2.2017
OS: win32 ia32
Adapter node: v6.5.0 ia32
vscode-chrome-debug-core: 3.12.8
debugger-for-chrome: 2.5.3
From client: initialize({"adapterID":"chrome","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true})
To client: {"seq":0,"type":"response","request_seq":1,"command":"initialize","success":true,"body":{"exceptionBreakpointFilters":[{"label":"All Exceptions","filter":"all","default":false},{"label":"Uncaught Exceptions","filter":"uncaught","default":true}],"supportsConfigurationDoneRequest":true,"supportsSetVariable":true,"supportsConditionalBreakpoints":true,"supportsCompletionsRequest":true,"supportsHitConditionalBreakpoints":true,"supportsRestartFrame":true,"supportsRestartRequest":true}}
From client: launch({"name":"Chrome with sourcemaps","type":"chrome","request":"launch","url":"http://localhost:8080/index.html","sourceMaps":true,"webRoot":"c:\\Dev\\LRA5\\LRA5-FrontEnd\\FrontEnd","verboseDiagnosticLogging":true,"diagnosticLogging":true})
spawn('C:\Program Files (x86)\Google\Chrome\Application\chrome.exe', ["--remote-debugging-port=9222","--no-first-run","--no-default-browser-check","http://localhost:8080/index.html"])
Discovering targets via http://127.0.0.1:9222/json
Attaching to target: {"description":"","devtoolsFrontendUrl":"/devtools/inspector.html?ws=127.0.0.1:9222/devtools/page/137770c0-6895-4e3d-b8b0-22e98efc915f","id":"137770c0-6895-4e3d-b8b0-22e98efc915f","title":"ifm","type":"page","url":"http://localhost:8080/index.html","webSocketDebuggerUrl":"ws://127.0.0.1:9222/devtools/page/137770c0-6895-4e3d-b8b0-22e98efc915f"}
....... // Had to skip much, as I cannot post that much
From client: variables({"variablesReference":100066})
To client: {"seq":0,"type":"response","request_seq":72,"command":"variables","success":true,"body":{"variables":[{"name":"0","value":"Activating element undefined","variablesReference":0,"evaluateName":"0"}]}}
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"79","url":"","startLine":0,"startColumn":0,"endLine":324,"endColumn":61,"executionContextId":2,"hash":"4F409970DBBE83A872ACFE60FA07F6A97C99BF42","executionContextAuxData":{"isDefault":true,"frameId":"20232.1"},"isLiveEdit":false,"sourceMapURL":"controller/hosts/Hosts.controller.js.map","hasSourceURL":false}}
SourceMaps.getMapForGeneratedPath: Finding SourceMap for eval://79 by URI: controller/hosts/Hosts.controller.js.map and webRoot: c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"80","url":"","startLine":0,"startColumn":0,"endLine":141,"endColumn":66,"executionContextId":2,"hash":"50A3E723AB380B2676B9B50427E36A3838CB4E56","executionContextAuxData":{"isDefault":true,"frameId":"20232.1"},"isLiveEdit":false,"sourceMapURL":"controller/hosts/HostDetail.controller.js.map","hasSourceURL":false}}
SourceMaps.getMapForGeneratedPath: Finding SourceMap for eval://80 by URI: controller/hosts/HostDetail.controller.js.map and webRoot: c:\Dev\LRA5\LRA5-FrontEnd\FrontEnd
From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"log","text":"Initializing HostDetailController","url":"","line":84,"column":45}}}
From target: {"method":"Runtime.consoleAPICalled","params":{"type":"log","args":[{"type":"string","value":"Initializing HostDetailController"}],"executionContextId":2,"timestamp":1488210526767.364013671875000,"stackTrace":{"callFrames":[{"functionName":"HostDetail.onInit","scriptId":"80","url":"","lineNumber":83,"columnNumber":44},{"functionName":"a.fireEvent","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":447,"columnNumber":252},{"functionName":"a.fireEvent","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":998,"columnNumber":208},{"functionName":"","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":576,"columnNumber":285},{"functionName":"c._initCompositeSupport","scriptId":"55","url":"","lineNumber":905,"columnNumber":1157},{"functionName":"","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":459,"columnNumber":890},{"functionName":"constructor","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":459,"columnNumber":1013},{"functionName":"constructor","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":978,"columnNumber":28},{"functionName":"constructor","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":813,"columnNumber":58},{"functionName":"f","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":641,"columnNumber":423},{"functionName":"f","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":641,"columnNumber":423},{"functionName":"sap.ui.view","scriptId":"55","url":"","lineNumber":924,"columnNumber":874},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6307},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"g","scriptId":"55","url":"","lineNumber":569,"columnNumber":2496},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5484},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5203},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6171},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"g","scriptId":"55","url":"","lineNumber":569,"columnNumber":2496},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5484},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6171},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"g","scriptId":"55","url":"","lineNumber":569,"columnNumber":2496},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5484},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5203},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6171},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"a","scriptId":"55","url":"","lineNumber":569,"columnNumber":1330},{"functionName":"b","scriptId":"55","url":"","lineNumber":569,"columnNumber":1744},{"functionName":"X.parseTemplate","scriptId":"55","url":"","lineNumber":569,"columnNumber":615},{"functionName":"M.runWithPreprocessors.settings","scriptId":"55","url":"","lineNumber":952,"columnNumber":111},{"functionName":"h.runWithPreprocessors","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":465,"columnNumber":197},{"functionName":"f.onControllerConnected","scriptId":"55","url":"","lineNumber":952,"columnNumber":59},{"functionName":"I","scriptId":"55","url":"","lineNumber":905,"columnNumber":810},{"functionName":"r","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":662,"columnNumber":61},{"functionName":"h.runAsOwner","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":675,"columnNumber":42},{"functionName":"p","scriptId":"55","url":"","lineNumber":905,"columnNumber":568},{"functionName":"","scriptId":"55","url":"","lineNumber":905,"columnNumber":937}]}}}
From client: variables({"variablesReference":100067})
To client: {"seq":0,"type":"response","request_seq":73,"command":"variables","success":true,"body":{"variables":[{"name":"0","value":"Initializing HostDetailController","variablesReference":0,"evaluateName":"0"}]}}
From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"debug","text":"HostDetailController is subscribing to event host.selected","url":"http://localhost:8080/scripts/baseEvent.js","line":85,"column":45}}}
From target: {"method":"Runtime.consoleAPICalled","params":{"type":"debug","args":[{"type":"string","value":"HostDetailController is subscribing to event host.selected"}],"executionContextId":2,"timestamp":1488210526788.474121093750000,"stackTrace":{"callFrames":[{"functionName":"Event.subscribe","scriptId":"60","url":"http://localhost:8080/scripts/baseEvent.js","lineNumber":84,"columnNumber":44},{"functionName":"HostDetail.onInit","scriptId":"80","url":"","lineNumber":90,"columnNumber":69},{"functionName":"a.fireEvent","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":447,"columnNumber":252},{"functionName":"a.fireEvent","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":998,"columnNumber":208},{"functionName":"","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":576,"columnNumber":285},{"functionName":"c._initCompositeSupport","scriptId":"55","url":"","lineNumber":905,"columnNumber":1157},{"functionName":"","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":459,"columnNumber":890},{"functionName":"constructor","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":459,"columnNumber":1013},{"functionName":"constructor","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":978,"columnNumber":28},{"functionName":"constructor","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":813,"columnNumber":58},{"functionName":"f","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":641,"columnNumber":423},{"functionName":"f","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":641,"columnNumber":423},{"functionName":"sap.ui.view","scriptId":"55","url":"","lineNumber":924,"columnNumber":874},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6307},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"g","scriptId":"55","url":"","lineNumber":569,"columnNumber":2496},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5484},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5203},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6171},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"g","scriptId":"55","url":"","lineNumber":569,"columnNumber":2496},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5484},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6171},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"g","scriptId":"55","url":"","lineNumber":569,"columnNumber":2496},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5484},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"Q","scriptId":"55","url":"","lineNumber":569,"columnNumber":5203},{"functionName":"P","scriptId":"55","url":"","lineNumber":569,"columnNumber":5100},{"functionName":"k","scriptId":"55","url":"","lineNumber":569,"columnNumber":6171},{"functionName":"h","scriptId":"55","url":"","lineNumber":569,"columnNumber":2753},{"functionName":"a","scriptId":"55","url":"","lineNumber":569,"columnNumber":1330},{"functionName":"b","scriptId":"55","url":"","lineNumber":569,"columnNumber":1744},{"functionName":"X.parseTemplate","scriptId":"55","url":"","lineNumber":569,"columnNumber":615},{"functionName":"M.runWithPreprocessors.settings","scriptId":"55","url":"","lineNumber":952,"columnNumber":111},{"functionName":"h.runWithPreprocessors","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":465,"columnNumber":197},{"functionName":"f.onControllerConnected","scriptId":"55","url":"","lineNumber":952,"columnNumber":59},{"functionName":"I","scriptId":"55","url":"","lineNumber":905,"columnNumber":810},{"functionName":"r","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":662,"columnNumber":61},{"functionName":"h.runAsOwner","scriptId":"53","url":"http://localhost:8080/resources/sap-ui-core.js","lineNumber":675,"columnNumber":42},{"functionName":"p","scriptId":"55","url":"","lineNumber":905,"columnNumber":568},{"functionName":"","scriptId":"55","url":"","lineNumber":905,"columnNumber":937}]}}}
From client: variables({"variablesReference":100068})
To client: {"seq":0,"type":"response","request_seq":74,"command":"variables","success":true,"body":{"variables":[{"name":"0","value":"HostDetailController is subscribing to event host.selected","variablesReference":0,"evaluateName":"0"}]}}
..... // Skipped rest of the log.

```

**Comment by roblourens** [comment]
I see the problem. The sourcemap URL usually gets resolved relative to the path of the script, but when the script is eval'd it loses its path, so the debugger can't find the sourcemap file.

If you change it to write out sourcemap paths with a leading slash (`/controller/hosts/Hosts.controller.js.map`) then I think it might work as-is. I will also make it try to resolve webRoot + sourcemap path as a fallback, which is probably a good thing to do.

**Comment by apazureck** [comment]
Ok, that sounds great. Could you give me a little hint how to change that? Can I do that using the sourcemappathoverrides? I tried to set the paths to the files earlier, but this did not help.

**Comment by apazureck** [comment]
Hey there,
I played a little around with gulp, browsersync, etc. and had some success, but also a weird behavior I cannot explain.

First of all. If my js file is on my root level of my webservice, the breakpoint gets hit even without adding the script file to my index.html.

But if I want to get a file located in a subdirectory (for example controller) the source map resolution adds an additional folder up to the source map path.

Here is my launch.json:

```json
{
    "version": "0.2.0",
    "configurations": [{
        "type": "chrome",
        "request": "launch",
        "name": "Launch Chrome",
        "url": "http://localhost:3000",
        "webRoot": "${workspaceRoot}/out",
        "sourceMaps": true,
        "disableNetworkCache": true,
        "sourceMapPathOverrides": {
            "*": "${workspaceRoot}/src/*"
        }
    }]
}
```

And I get this using `.scripts` in the debug console:

```
http://localhost:3000/Component.js?eval (c:\Dev Env\UI5TypescriptDeclarations\generator\out\Component.js)
    - ..\src\Component.ts (c:\Dev Env\UI5TypescriptDeclarations\generator\src\Component.ts)
http://localhost:3000/controller/App.controller.js?eval (c:\Dev Env\UI5TypescriptDeclarations\generator\out\controller\App.controller.js)
    - ..\..\..\src\controller\App.controller.ts (c:\Dev Env\src\controller\App.controller.ts)
```

As you can see in the output, my Component.ts is resolved correctly. Go one folder up, go down to source. My breakpoint gets hit if set in Component.ts.

The App.controller.ts gets one additional ../ for some reason. So it is one folder too high and not finding the source file. How can this be? Can I fix this somehow, or am I using the sourcemap path overrides wrong here?

Here is the link to the repo:
https://github.com/apazureck/UI5TypescriptDeclarations/tree/master/generator

Thank you for your help and the great work you are doing.

**Comment by roblourens** [comment]
I've had similar problems with gulp-sourcemaps, and it takes a lot of fiddling to get it to do the right thing. If you look at your .js.map file for 'controller', it has `sourceRoot: '../..'` and `sources: ['../src/controller/...']`. No idea why it wants to write the sources entry relative to 'out', but you might have to take that into consideration when calculating sourceRoot.

**Comment by apazureck** [comment]
OK, thank you for that hint. I would have never thought that may be the source of my problem. I will look into that.

**Comment by apazureck** [comment]
Hi again @roblourens,
I tried to use the source maps provided by the ts compiler.

`return exec("tsc");`

It even points to the source files:
`{"version":3,"file":"Component.js","sourceRoot":"c:/Dev Env/UI5TypescriptDeclarations/generator/src/","sources":["Component.ts"]`
and
`{"version":3,"file":"App.controller.js","sourceRoot":"c:/Dev Env/UI5TypescriptDeclarations/generator/src/","sources":["controller/App.controller.ts"]`

But if I do it this way the chrome debugger does not load the scripts anymore (used .scripts). So relative sourcemaps seem to be the way to go here?

BUT, I found a solution. Seems like gulp-sourcemaps automatically adds a relative path if you provide the sourcemap option, so I removed the relative path calculation and just return an empty string. Works like a charm. Thank's for your help.
